### PR TITLE
update verify validations

### LINKED ISSUE #128: Update validation for verify function in SMT
### Description

Additional validation are added on SMT verify method on https://github.com/LiskHQ/lips/pull/431. 

Specifically, validation below are added
```
# If the value is empty, both queries point to an empty node.
# The binary path must match (the keys can be different).
if q.value == EMPTY_VALUE:
    if q.binary_path != duplicateQuery.binary_path:
        return False
# Otherwise, the keys must be the same.
else:
    if q.key != duplicateQuery.key:
        return False
````

### Acceptance Criteria

- All existing tests must pass 
- Add additional unit test to ensure new validation is correct

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #128 

### How was it solved?
 - Add new verification in the verify function of the SMT
 - Also refactor codes to combine two functions into one to reduce the number of the loops over the tree.

### How was it tested?
 - Add a new unit test for new checks.
 - All tests passed

### CODE REVIEW

**Review by matjazv — APPROVED** [review]
Should we add some additional unit tests to check newly implemented checks/validations?

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: src/sparse_merkle_tree/smt.rs [hunk 1]
```diff
@@ -842,38 +842,74 @@ impl SparseMerkleTree {
...
 
-    fn verify_query_keys(proof: &Proof, query_keys: &[Vec<u8>], key_length: KeyLength) -> bool {
-        let mut queries: HashMap<&[u8], QueryProof> = HashMap::new();
+    /// verification_and_prepare_with_proof_map checks all verifications of query_keys based on the verify function in the [LIP-0039](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0039.md#proof-construction).
+    /// Also it returns a tuple, the first parameter is the result of the query_keys verification and the second one is a map
+    fn verification_and_prepare_with_proof_map(
+        proof: &Proof,
+        query_keys: &[Vec<u8>],
+        key_length: KeyLength,
+    ) -> Result<(bool, HashMap<Vec<bool>, QueryProofWithProof>), SMTError> {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:851]: I think if we have other types of error will be better but I tired to implement new function like the old one, so if you agree I can change it.

File: src/sparse_merkle_tree/smt.rs [hunk 2]
```diff
@@ -842,38 +842,74 @@ impl SparseMerkleTree {
...
-                if !utils::is_bytes_equal(&q.bitmap, &query.bitmap)
-                    || !utils::is_bytes_equal(q.value(), query.value())
+            let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
+            let query_key_binary = utils::bytes_to_bools(query.key());
+            let binary_path = if binary_bitmap.len() > query_key_binary.len() {
+                return Err(SMTError::InvalidBitmapLen);
+            } else {
+                query_key_binary[..binary_bitmap.len()].to_vec()
+            };
+            if let Some(duplicate_query) = queries.get(&binary_path) {
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:875]: I tried to use it but test cases will fail with this one.

File: src/sparse_merkle_tree/smt.rs [hunk 3]
```diff
@@ -842,38 +842,74 @@ impl SparseMerkleTree {
         Ok((query_with_proofs, ancestor_hashes))
     }
 
-    fn verify_query_keys(proof: &Proof, query_keys: &[Vec<u8>], key_length: KeyLength) -> bool {
-        let mut queries: HashMap<&[u8], QueryProof> = HashMap::new();
+    /// verification_and_prepare_with_proof_map checks all verifications of query_keys based on the verify function in the [LIP-0039](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0039.md#proof-construction).
+    /// Also it returns a tuple, the first parameter is the result of the query_keys verification and the second one is a map
+    fn verification_and_prepare_with_proof_map(
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:847]: Done.

File: src/sparse_merkle_tree/smt.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shuse2) [reply src/sparse_merkle_tree/smt.rs:851]: yea i think it's better to keep the interface simple

**Comment by hrmhatef** [comment]
> Should we add some additional unit tests to check newly implemented checks/validations?

The new unitest `test_verify` will check new validations.

File: src/sparse_merkle_tree/smt.rs [hunk 4]
```diff
@@ -2538,13 +2887,12 @@ mod tests {
             1u8, 2, 3, 4, 1, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71,
             3, 3, 71, 3, 3, 71, 3,
         ]);
-        let res = SparseMerkleTree::verify(
+        let res = SparseMerkleTree::verify_and_prepare_proof_map(
```
Comment (matjazv) [inline src/sparse_merkle_tree/smt.rs:2890]: Why is this change needed in this unit test and not also in others (for example in `test_mid_proof()`)?

**Review by matjazv — CHANGES_REQUESTED** [review]

**Comment by matjazv** [comment]
> > Should we add some additional unit tests to check newly implemented checks/validations?
> 
> The new unitest `test_verify` will check new validations.

Should we expect some more commits based on that in `test_verify` only a happy path is tested?

**Comment by hrmhatef** [comment]
> > > Should we add some additional unit tests to check newly implemented checks/validations?
> > 
> > 
> > The new unitest `test_verify` will check new validations.
> 
> Should we expect some more commits based on that in `test_verify` only a happy path is tested?

Yes it will check the happy path, you mean we need to add more tests for that?

**Comment by matjazv** [comment]
> > > > Should we add some additional unit tests to check newly implemented checks/validations?
> > > 
> > > 
> > > The new unitest `test_verify` will check new validations.
> > 
> > 
> > Should we expect some more commits based on that in `test_verify` only a happy path is tested?
> 
> Yes it will check the happy path, you mean we need to add more tests for that?

I'm asking if it's needed.
If newly added validations are critical/important then it would be nice to actually test them. If not, then new unit tests are not needed.

File: src/sparse_merkle_tree/smt.rs [hunk 5]
```diff
@@ -842,38 +842,78 @@ impl SparseMerkleTree {
...
-    fn verify_query_keys(proof: &Proof, query_keys: &[Vec<u8>], key_length: KeyLength) -> bool {
-        let mut queries: HashMap<&[u8], QueryProof> = HashMap::new();
+    /// verify_and_prepare_proof_map checks all verifications of query_keys based on the verify function in the [LIP-0039](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0039.md#proof-construction).
+    /// Also it returns a tuple, the first parameter is the result of the query_keys verification and the second one is a map
+    fn verify_and_prepare_proof_map(
+        proof: &Proof,
+        query_keys: &[Vec<u8>],
+        key_length: KeyLength,
+    ) -> Result<HashMap<Vec<bool>, QueryProofWithProof>, SMTError> {
+        let mut queries_with_proof: HashMap<Vec<bool>, QueryProofWithProof> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:852]: it will initialize by `HashMap::new()`, you mean, I can initialize it by specific data?

[the remark this answers is not in the document]
Comment (matjazv) [reply src/sparse_merkle_tree/smt.rs:852]: @ricott1 probably meant that line 852 could be moved down to line 859. Speed optimization, unnecessary initialization of hash map in case that if statement returns from function.

File: src/sparse_merkle_tree/smt.rs [hunk 6]
```diff
@@ -842,38 +842,78 @@ impl SparseMerkleTree {
...
             }
+
             let query = &proof.queries[i];
-            let duplicate_query = queries.get(query.key());
-            if let Some(q) = duplicate_query {
-                if !utils::is_bytes_equal(&q.bitmap, &query.bitmap)
-                    || !utils::is_bytes_equal(q.value(), query.value())
+            let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
+            let query_key_binary = utils::bytes_to_bools(query.key());
+            let binary_path = if binary_bitmap.len() > query_key_binary.len() {
```
Comment (ricott1) [inline src/sparse_merkle_tree/smt.rs:872]: I think this check should be done only if the query key and the relative proof query key are unequal, just like indicated in the LIP

File: src/sparse_merkle_tree/smt.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:852]: Done, sorry for my misunderstanding.

**Review by matjazv — APPROVED** [review]

File: src/sparse_merkle_tree/smt.rs [hunk 7]
```diff
@@ -1537,25 +1572,11 @@ impl SparseMerkleTree {
...
-        let filter_map = Self::prepare_queries_with_proof_map(proof)?;
-        let mut filtered_proof = filter_map
-            .values()
-            .cloned()
-            .collect::<Vec<QueryProofWithProof>>();
+        let mut filtered_proof =
+            match Self::verify_and_prepare_proof_map(proof, query_keys, key_length) {
+                Ok(v) => v,
+                Err(_) => return Ok(false),
+            };
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:1579]: I guess we can not use `unwarp_or` error here because we want to return a `bool` instead of `QueryProofWithProof`.

[the remark this answers is not in the document]
Comment (shuse2) [reply src/sparse_merkle_tree/smt.rs:1579]: ah true 👍

**Review by ricott1 — APPROVED** [review]
